SignalMessaging/profiles/OWSProfileManager.m

```diff
@@ -1029,131 +768,123 @@ - (nullable UIImage *)profileAvatarForRecipientId:(NSString *)recipientId
+        OWSAES256Key *profileKeyAtStart = userProfile.profileKey;
 
-            NSString *fileName = [[NSUUID UUID].UUIDString stringByAppendingPathExtension:@"jpg"];
-            NSString *filePath = [self.profileAvatarsDirPath stringByAppendingPathComponent:fileName];
+        NSString *fileName = [[NSUUID UUID].UUIDString stringByAppendingPathExtension:@"jpg"];
+        NSString *filePath = [self.profileAvatarsDirPath stringByAppendingPathComponent:fileName];
 
+        @synchronized(self.currentAvatarDownloads)
+        {
             if ([self.currentAvatarDownloads containsObject:userProfile.recipientId]) {
                 // Download already in flight; ignore.
                 return;
             }
             [self.currentAvatarDownloads addObject:userProfile.recipientId];
+        }
 
-            NSString *tempDirectory = NSTemporaryDirectory();
-            NSString *tempFilePath = [tempDirectory stringByAppendingPathComponent:fileName];
-
-            NSURL *avatarUrlPath =
-                [NSURL URLWithString:userProfile.avatarUrlPath relativeToURL:self.avatarHTTPManager.baseURL];
-            NSURLRequest *request = [NSURLRequest requestWithURL:avatarUrlPath];
-            NSURLSessionDownloadTask *downloadTask = [self.avatarHTTPManager downloadTaskWithRequest:request
-                progress:^(NSProgress *_Nonnull downloadProgress) {
-                    DDLogVerbose(@"%@ Downloading avatar for %@ %f",
-                        self.logTag,
-                        userProfile.recipientId,
-                        downloadProgress.fractionCompleted);
+        NSString *tempDirectory = NSTemporaryDirectory();
+        NSString *tempFilePath = [tempDirectory stringByAppendingPathComponent:fileName];
+
+        void (^completionHandler)(NSURLResponse *_Nonnull, NSURL *_Nullable, NSError *_Nullable) = ^(
+            NSURLResponse *_Nonnull response, NSURL *_Nullable filePathParam, NSError *_Nullable error) {
+            // Ensure disk IO and decryption occurs off the main thread.
+            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+                NSData *_Nullable encryptedData = (error ? nil : [NSData dataWithContentsOfFile:tempFilePath]);
+                NSData *_Nullable decryptedData = [self decryptProfileData:encryptedData profileKey:profileKeyAtStart];
+                UIImage *_Nullable image = nil;
+                if (decryptedData) {
+                    BOOL success = [decryptedData writeToFile:filePath atomically:YES];
+                    if (success) {
+                        image = [UIImage imageWithContentsOfFile:filePath];
+                    }
                 }
-                destination:^NSURL *_Nonnull(NSURL *_Nonnull targetPath, NSURLResponse *_Nonnull response) {
-                    return [NSURL fileURLWithPath:tempFilePath];
+
+                @synchronized(self.currentAvatarDownloads)
```

Comment:
There's still a couple of properties we need to synchronize access to - but we now do that narrowly and sync on those objects themselves which is safer.